Add `verify_base_db` method to check if all base tables are present.

vedanshujain opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Adding another flag `woocommerce-schema-version`.

One of the notable changes in this PR is that we are creating another flag: `woocommerce-schema-version` to store the version till which we have verified that base schema DB is installed.

Even though we already have similar flags:

1. `woocommerce_version` to store the WooCommerce Version

2. `woocommerce_db_version` to store the version point till where we have run the update function.

I wanted to use `woocommerce_db_version` but it was not suitable because having not run update callbacks is quite different from base schema structure mismatch. Also, I did not want to prevent update callback here because this responsibility of erroring and scheduling themselves again should fall to the callback code itself.

### Changes proposed in this Pull Request:

 Add `verify_base_db` method to check if all base tables are present.

Optionally, it also adds a notice in case all DB tables are not present.

Note that we only check missing tables and don't care about exact table structure because many time tables are modified by merchants to better suit their needs (indexes, collations, etc).

### How to test the changes in this Pull Request:

Part 1

1. Rename a base table, maybe `{prefix}wc_tax_rate_classes` to `wc_tax_rate_classes_2`.
1. Comment out `self::create_tables();` line to make sure table is not created on refresh.
1. Delete option `woocommerce_version` and `woocommerce_db_version` to trigger update flow.
1. You should be able to see notice similar to 
``` 
One or more tables required for WooCommerce to functions are missing, some features may not work as expected. Missing tables: wc_wc_tax_rate_classes. 
```
Don't dismiss the notice just yet.
1. Apply the patch in https://github.com/woocommerce/woocommerce-rest-api/pull/188
1. Rename the table back to `{prefix}wc_tax_rate_classes`.
1. Refresh the page.
1. Click on "Check again" text in the notice.
1. Notice will go away on refresh.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

<!-- Mark completed items with an [x] -->

### Changelog entry

> Dev - Add `verify_base_db` method to check if all base tables are present.

vedanshujain committed:
Add `verify_base_db` method to check if all base tables are present.

Optionally, also adds a notice in case all db tables are not present. Returns list of tables.

Note that we only check missing tables and don't care about exact table structure because many time tables are modified by merchants to better suit their needs (indexes, collations etc).

vedanshujain committed:
Add support for notices

vedanshujain committed:
Modified notice to also handle when REST API does not have verify tool.

vedanshujain committed:
Add unit test for verify_base_tables function

Konamiman commented:
Before applying the REST API patch:

![image](https://user-images.githubusercontent.com/937723/81823956-326c8700-9535-11ea-845b-6ce6eb654da3.png)

After applying the patch:

![image](https://user-images.githubusercontent.com/937723/81824632-fe459600-9535-11ea-9cdb-22291cb5c01b.png)

After renaming back the table, clicking "Check again" and refreshing:

![image](https://user-images.githubusercontent.com/937723/81824869-3c42ba00-9536-11ea-970c-44f5dc5afcb9.png)

So 👍🏻

vedanshujain committed:
Better option names to properly describe their intention.

Additionally, allow an option in `verify_base_tables` to call `create_table` method. This will be useful if someone wants to fix the their database.

vedanshujain committed:
Add test to ensure that verify_base_table also creates table if needed.

vedanshujain committed:
Changed back method to private because we figured a workaround in earlier test

claudiosanches commented on includes/admin/views/html-notice-base-table-missing.php:
```diff
@@ -0,0 +1,44 @@
+				/* translators: %1%s: Missing tables (seperated by ",") %2$s: Link to check again */
+					__( 'One or more tables required for WooCommerce to function are missing, some features may not work as expected. Missing tables: %1$s. <a href="%2$s">Check again.</a>', 'woocommerce' ),
+					esc_html( implode( ', ', $missing_tables ) ),
+					wp_nonce_url( admin_url( 'admin.php?page=wc-status&tab=tools&action=verify_db_tables' ), 'debug_action' )
```
Missing `esc_url`.

vedanshujain replied:
Changed in https://github.com/woocommerce/woocommerce/pull/26454/commits/4fce106f1446eb5e8fec45bd040b2cacde971880

claudiosanches replied:
WordPress mention that the URL get escaped, but in several places also uses `esc_url()`, on WooCommerce we have the same. But sounds like we don't need to worry about it here.

claudiosanches commented on includes/admin/views/html-notice-base-table-missing.php:
```diff
@@ -0,0 +1,44 @@
+	<p>
+		<?php
+		$verify_db_tool_available = array_key_exists( 'verify_db_tables', WC_Admin_Status::get_tools() );
+		$missing_tables           = get_option( 'woocommerce-schema-missing-tables' );
```
It need to get updated to `woocommerce-schema-missing-tables`, by default on WooCommerce we use `woocommerce_` as a prefix for all options, note that also we remove options with this prefix when running the `unistall.php`.

vedanshujain replied:
Changed in https://github.com/woocommerce/woocommerce/pull/26454/commits/4fce106f1446eb5e8fec45bd040b2cacde971880

claudiosanches commented on includes/class-wc-install.php:
```diff
@@ -303,6 +304,43 @@ public static function install() {
+			if ( $modify_notice ) {
+				WC_Admin_Notices::remove_notice( 'base_tables_missing' );
+			}
+			update_option( 'woocommerce-schema-version', WC()->version );
```
Need to replace `-` for `_` in this option key.

vedanshujain replied:
Changed in https://github.com/woocommerce/woocommerce/pull/26454/commits/4fce106f1446eb5e8fec45bd040b2cacde971880

claudiosanches reviewed (changes requested):
Looks good, great work and makes sense to save the database schema in a new option.

I placed some notes about tweaks that we could do, but I think we could have a database schema version instead of using WooCommerce's version, now we are registering the WooCommerce version used when the user updated and had the database version updated, but this isn't the database schema version.

claudiosanches committed:
Fixed coding standards

vedanshujain committed:
Use `_` instead of `-` as per convention.

vedanshujain committed:
Use seperate db version to better track changes

claudiosanches reviewed (approved):
Great work!
